### PR TITLE
Add ChemoMetic NucleoView parser

### PR DESCRIPTION
**Pull request by bwolfe-benchling**
This PR adds support for the ChemoMetec NucleoView cell counting software.

### CODE REVIEW

File: src/allotropy/parsers/chemometec_nucleoview/nucleoview_reader.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+def convert_float(x: pd.Series) -> pd.Series:
+    return pd.to_numeric(x, errors="coerce")
+
+
+def convert_int(x: pd.Series) -> pd.Series:
+    return pd.to_numeric(x, errors="coerce")
+
+
+def convert_string(x: pd.Series) -> pd.Series:
```
Comment (bdworth) [inline src/allotropy/parsers/chemometec_nucleoview/nucleoview_reader.py:19]: Unless you intend these functions to be called outside of this package, could you please "protect" them? E.g., `_convert_string`.

**Review by bdworth — CHANGES_REQUESTED** [review]
Thanks for contributing! I left some initial comments. Also, note that build checks fail fast here, so the current failure due to unsorted imports might be masking other lint/test issues.

Also overall comment: what is the plan to test these additions?

File: src/allotropy/parsers/chemometec_nucleoview/nucleoview_parser.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+            return str(data["Instrument type"].iloc[0])
+        except KeyError:
+            return DEFAULT_MODEL_NUMBER
+
+    def _get_cell_counting_document(
+        self, data: pd.DataFrame
+    ) -> list[CellCountingDocumentItem]:
+        return [
+            self._get_cell_counting_document_item(sample)
+            for _, sample in data.iterrows()
```
Comment (bdworth) [inline src/allotropy/parsers/chemometec_nucleoview/nucleoview_parser.py:112]: oops, I missed this earlier. `iterrows` is a big pandas antipattern. I now know why you did this after looking at other vi-cell code (and fixing it!). Could you please refactor a la #144?

File: src/allotropy/parsers/chemometec_nucleoview/nucleoview_parser.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+def get_property_from_sample(sample: pd.Series, property_name: str) -> Any:
+    if (value := sample.get(property_name)) is not None:
+        property_type = property_lookup[property_name]
+
+        # if the porperty type is measured in million cells per ml convert cells per ml
+        if property_type == TQuantityValueMillionCellsPerMilliliter:
+            return property_type(value=float(str(value)) / 1e6)
+
+        return property_type(value=value)  # type: ignore[arg-type]
```
[the remark this answers is not in the document]
Comment (bdworth) [reply src/allotropy/parsers/chemometec_nucleoview/nucleoview_parser.py:58]: Some of the type hints could be fixed using proper typing, a la #139

However, that masks a bigger issue: we shouldn't ever be using `iterrows` to get a row as a `Series`. That said, some suppressions may be necessary based on how we're extracting properties in `get_property`. See #144 for examples of both.

**Review by bdworth — CHANGES_REQUESTED** [review]

File: src/allotropy/parsers/chemometec_nucleoview/nucleoview_parser.py (hunk 2 above)
Comment (bwolfe-benchling) [reply src/allotropy/parsers/chemometec_nucleoview/nucleoview_parser.py]: latest commit refactors to mimic 144...I think

File: src/allotropy/parsers/chemometec_nucleoview/nucleoview_reader.py (hunk 1 above)
Comment (bwolfe-benchling) [reply src/allotropy/parsers/chemometec_nucleoview/nucleoview_reader.py]: removed from latest refactor

File: src/allotropy/parsers/chemometec_nucleoview/nucleoview_reader.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    return x.astype(str)
+
+
+conversors = {
```
[the remark this answers is not in the document]
Comment (bwolfe-benchling) [reply src/allotropy/parsers/chemometec_nucleoview/nucleoview_reader.py]: removed from latest refactor

File: src/allotropy/parsers/chemometec_nucleoview/nucleoview_parser.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    ) -> Optional[str]:
+        try:
+            return str(data["Instrument s/n"].iloc[0])
+        except KeyError:
+            return None
+
+    def _get_device_identifier(
+        self,
+        data: pd.DataFrame,
+    ) -> Optional[str]:
```
[the remark this answers is not in the document]
Comment (bwolfe-benchling) [reply src/allotropy/parsers/chemometec_nucleoview/nucleoview_parser.py]: removed in latest refactor

File: src/allotropy/parsers/chemometec_nucleoview/nucleoview_parser.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+from allotropy.constants import ASM_CONVERTER_NAME, ASM_CONVERTER_VERSION
+from allotropy.parsers.chemometec_nucleoview.constants import (
+    DEFAULT_ANALYST,
+    DEFAULT_MODEL_NUMBER,
+    NUCLEOCOUNTER_SOFTWARE_NAME,
+)
+from allotropy.parsers.chemometec_nucleoview.nucleoview_reader import NC200Reader
+from allotropy.parsers.vendor_parser import VendorParser
+
+property_lookup = {
```
[the remark this answers is not in the document]
Comment (bwolfe-benchling) [reply src/allotropy/parsers/chemometec_nucleoview/nucleoview_parser.py]: done

File: src/allotropy/parser_factory.py [hunk 7]
```diff
@@ -29,10 +32,11 @@
 
 class Vendor(Enum):
     AGILENT_GEN5 = "AGILENT_GEN5"
-    APPBIO_QUANTSTUDIO = "APPBIO_QUANTSTUDIO"
     APPBIO_ABSOLUTE_Q = "APPBIO_ABSOLUTE_Q"
+    APPBIO_QUANTSTUDIO = "APPBIO_QUANTSTUDIO"
```
Comment (bdworth) [inline src/allotropy/parser_factory.py:36]: Thanks for doing this, but unrelated cleanup should really be in a separate PR. No need to remove -- I'm doing here: #155

File: tests/parsers/chemometec_nucleoview/chemometec_nucleoview_parser_test.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
+import pytest
+
+from allotropy.parser_factory import Vendor
+from tests.parsers.test_utils import from_file, validate_schema
+
+output_files = (
```
Comment (bdworth) [inline tests/parsers/chemometec_nucleoview/chemometec_nucleoview_parser_test.py:6]: `OUTPUT_FILES`

File: tests/parsers/chemometec_nucleoview/chemometec_nucleoview_parser_test.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    "chemometec_nucleoview_example01",
+    "chemometec_nucleoview_example02",
+    "chemometec_nucleoview_example03",
```
Comment (bdworth) [inline tests/parsers/chemometec_nucleoview/chemometec_nucleoview_parser_test.py:9]: could you please put the .csv extensions in here?

File: src/allotropy/parsers/chemometec_nucleoview/nucleoview_reader.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+                ~raw_data["Time zone offset"].str.startswith("-"), "Time zone offset"
+            ] = ("+" + raw_data["Time zone offset"])
+
+            # combine date time and UTC offset
+            raw_data["datetime"] = pd.to_datetime(
+                raw_data["Date time"] + raw_data["Time zone offset"]
+            )
+        raw_data["Sample ID"] = raw_data["Image"].str.split("-", n=3).str[3]
+
+        return cast(pd.DataFrame, raw_data)
```
Comment (bdworth) [inline src/allotropy/parsers/chemometec_nucleoview/nucleoview_reader.py:37]: rather than cast here, could you try casting when assigning to `df`? I think the reason pandas is confused is because `read_csv` can return `DataFrame | TextFileReader`.

**Review by bdworth — CHANGES_REQUESTED** [review]
thanks for the quick changes! Some more comments, still not a full review though.

File: src/allotropy/parser_factory.py (hunk 7 above)
Comment (bwolfe-benchling) [reply src/allotropy/parser_factory.py:36]: tbh, I didn't realize I did this.

File: src/allotropy/parsers/chemometec_nucleoview/nucleoview_reader.py (hunk 10 above)
Comment (bwolfe-benchling) [reply src/allotropy/parsers/chemometec_nucleoview/nucleoview_reader.py]: moved cast and had to adjust pivot for typing reasons

File: tests/parsers/chemometec_nucleoview/chemometec_nucleoview_parser_test.py (hunk 8 above)
Comment (bwolfe-benchling) [reply tests/parsers/chemometec_nucleoview/chemometec_nucleoview_parser_test.py]: done

File: tests/parsers/chemometec_nucleoview/chemometec_nucleoview_parser_test.py (hunk 9 above)
Comment (bwolfe-benchling) [reply tests/parsers/chemometec_nucleoview/chemometec_nucleoview_parser_test.py]: done

File: src/allotropy/parsers/chemometec_nucleoview/nucleoview_reader.py [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    @classmethod
+    def read(cls, contents: io.IOBase) -> pd.DataFrame:
+        df = cast(pd.DataFrame, pd.read_csv(contents, skipfooter=1, sep=";", index_col=False))  # type: ignore[call-overload]
+
+        # drop last column
+        df = df.drop(df.columns[2], axis=1)
+
+        # drop NA and blank values in our key column
+        df = df[df[df.columns[1]].notna()]
+        df = df[df[df.columns[1]] != " "]
```
Comment (bdworth) [inline src/allotropy/parsers/chemometec_nucleoview/nucleoview_reader.py:17]: I don't want to spend too much time on this, but I think you can use more built-in functionality to `read_csv` and avoid extra post-processing. E.g., something like:
`df = pd.read_csv(contents, skiprows=1, sep=";", usecols=[0, 1]).dropna()`

File: src/allotropy/parsers/chemometec_nucleoview/nucleoview_parser.py [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+    if value is None:
+        return None
+
+    property_type = _PROPERTY_LOOKUP[property_name]
+
+    # if the porperty type is measured in million cells per ml convert cells per ml
+    if property_type == TQuantityValueMillionCellsPerMilliliter:
+        return property_type(value=float(value) / 1e6)
+
+    return property_type(value=float(value))
```
Comment (bdworth) [inline src/allotropy/parsers/chemometec_nucleoview/nucleoview_parser.py:69]: is the float cast necessary here?

**Review by bdworth — CHANGES_REQUESTED** [review]
I think these are my last change requests, but I'll take another look tomorrow.

File: src/allotropy/parsers/chemometec_nucleoview/nucleoview_parser.py (hunk 12 above)
Comment (bwolfe-benchling) [reply src/allotropy/parsers/chemometec_nucleoview/nucleoview_parser.py:69]: yes, when I leave the cast out I end up with float values represented as strings in the final ASM output.  I think this is because I end up with a 2 column dataframe after pivot that roughly emulates a map or Series.  Since some of the rows have text in the "value" column I get text instead of floats.

File: src/allotropy/parsers/chemometec_nucleoview/nucleoview_reader.py (hunk 11 above)
Comment (bwolfe-benchling) [reply src/allotropy/parsers/chemometec_nucleoview/nucleoview_reader.py]: I adjusted.  I think it was a good suggestion.

**Review by bdworth — DISMISSED** [review]
Looks good! Using built-in functionality for `pd.read_csv` will improve maintainability. (I'd comment inline, but then the comment would have to be resolved before we can merge.)

Could you please add a PR description?

**Review by bdworth — DISMISSED** [review]

**Review by mdlu — DISMISSED** [review]

**Review by mdlu — DISMISSED** [review]

**Review by mdlu — APPROVED** [review]
